nx/lib/nx.ex

```
@@ -6408,15 +6408,151 @@ defmodule Nx do
+
+    {s1, s2} = Nx.Shape.dot(t1, c1, b1, t2, c2, b2)
+
+    {output_shape, output_names} = dot_batch_output_shape(s1, c1, names1, b1, s2, c2, names2, b2)
+
     out = %{t1 | type: output_type, names: output_names, shape: output_shape}
-    impl!(t1, t2).dot(out, t1, axes1, t2, axes2)
+    impl!(t1, t2).dot(out, t1, c1, b1, t2, c2, b2)
+  end
+
+  defp dot_batch_output_shape(s1, c1, names1, b1, s2, c2, names2, b2) do
+    {batch_dim, batch_name, s1, c1, names1, s2, c2, names2} =
+      prep_dot_batch_output(s1, c1, names1, b1, s2, c2, names2, b2)
+
+    # zip reduce without the batched dimensions
+    {output_shape, output_names} = Nx.Shape.zip_reduce(s1, c1, names1, s2, c2, names2)
+    # re-add the batched dimensions.
+    # TODO: make sure this is not skipping some sort of validation. e.g. duplicate names
+    if is_nil(batch_dim) do
+      {output_shape, output_names}
+    else
+      output_shape = Tuple.insert_at(output_shape, 0, batch_dim)
+      output_names = [batch_name | output_names]
+      {output_shape, output_names}
+    end
+  end
+
+  defp prep_dot_batch_output(s1, c1, names1, b1, s2, c2, names2, b2) do
+    case {b1, b2} do
+      {[], []} ->
+        {nil, nil, s1, c1, names1, s2, c2, names2}
+
+      {[0], [0]} ->
+        batch_dim = elem(s1, 0)
+        batch_name = hd(names1)
+        {s1, c1, names1} = shift_left_for_batch(s1, c1, names1)
+        {s2, c2, names2} = shift_left_for_batch(s2, c2, names2)
+        {batch_dim, batch_name, s1, c1, names1, s2, c2, names2}
+
+      {[0], []} ->
+        batch_dim = elem(s1, 0)
+        batch_name = hd(names1)
+        {s1, c1, names1} = shift_left_for_batch(s1, c1, names1)
+        {batch_dim, batch_name, s1, c1, names1, s2, c2, names2}
+
+      {[], [0]} ->
+        batch_dim = elem(s2, 0)
+        batch_name = hd(names2)
+        {s2, c2, names2} = shift_left_for_batch(s2, c2, names2)
+        {batch_dim, batch_name, s1, c1, names1, s2, c2, names2}
```

Should we raise a nice error message here? It seems only the highest dimension can be batched after all?

---

You're right. Actually `b1` and `b2` can only be of the same length as well, so these clauses don't make sense anymore

---

Nx.Shape.dot already raises in unsupported cases, so we can just not match on anything else here